### PR TITLE
Fix TFM inconsistency in BuildHost

### PR DESCRIPTION
**Pull request by jjonescz**
Should fix failures reported in https://github.com/dotnet/sdk/pull/45353#issuecomment-2531191550.

### CODE REVIEW

File: eng/targets/TargetFrameworks.props [hunk 1]
```diff
@@ -55,7 +55,7 @@
         <NetRoslyn>$(NetCurrent)</NetRoslyn>
         <NetRoslynSourceBuild>$(NetCurrent);$(NetPrevious)</NetRoslynSourceBuild>
         <NetRoslynAll>$(NetCurrent);$(NetPrevious)</NetRoslynAll>
-        <NetRoslynBuildHostNetCoreVersion>$(NetCurrent)</NetRoslynBuildHostNetCoreVersion>
+        <NetRoslynBuildHostNetCoreVersion>$(NetPrevious)</NetRoslynBuildHostNetCoreVersion>
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply eng/targets/TargetFrameworks.props:58]: deferring to @mthalman and @jaredpar who worked on this pattern.

[the remark this answers is not in the document]
Comment (mthalman) [reply eng/targets/TargetFrameworks.props:58]: cc @jasonmalinowski

[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: > Shouldn't roslyn's source-built product (built inside the VMR) only target NetCurrent?

Every time we've tried this in the past we've ended up breaking some other part of the build (there is always an older runtime or dependency hanging around that forces us into `$(NetPrevious)`). This is why we keep both current + prev even when in full source build. If you all are confident this no longer holds we can remove the all the `$(NetPrevious)` from this `<When>` block ... but I will small stakes bet we will find some way this breaks us again :wink:

Alternatively if we want to keep building current + previous I think the right fix is to change the project files so that they agree on TFM as they depend on each other: 

| Project | TFM Set |
| --- | ---|
| Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj | $(NetRoslynBuildHostNetCoreVersion);net472 | |
| Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj | $(NetRoslynSourceBuild);net472 |

Basically make the latter `$(NetRoslynBuildHostNetCoreVersion)`

[the remark this answers is not in the document]
Comment (jjonescz) [reply eng/targets/TargetFrameworks.props:58]: Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework here

https://github.com/dotnet/roslyn/blob/c8fba927cf53c6826e7dcf272abc1ff835a890f8/src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj#L101

so that will need to be updated somehow if NetRoslynBuildHostNetCoreVersion is changed to contain multiple values.

[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: > Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework (no plural) here

I've been thinking that we should change our approach to have our named TFs end with an 's' if they can be multiple in any configuration. That would make it easier to read project files at a glance.

[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply eng/targets/TargetFrameworks.props:58]: > Note that NetRoslynBuildHostNetCoreVersion is used as TargetFramework (no plural) here

I'm not sure why that variable isn't being used in the other location; looking through history I didn't see an obvious place where it was...might have been an oversight when the variable was originally introduced?

> if NetRoslynBuildHostNetCoreVersion is changed to contain multiple values

I can't imagine a case where we'd need multiple ones; it'd be tricky anyways to have to bundle multiple versions of the BuildHost and pick the right one as it is, so assuming there would only be a single TFM seems reasonable.

Generally, NetCurrent should be fine for the SDK build, other than Jared's concern that that often creates a surprise. For our official builds that produce our NuGet packages that go to nuget.org, "older is better" in terms of TFM, especially since tools like the upgrade assistants need to potentially analyze projects for upgrade and may not have newer runtimes on the box. (I'd seriously set it to netcoreapp3.1 if it wasn't likely to result in
image

[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: > Generally, NetCurrent should be fine for the SDK build, other than Jared's concern that that often creates a surprise.

The tricky part about the SDK is that Roslyn simultaneously flows to both .NET preview and .NET stable. Concretely at this moment every change in `main` will flow to both 9.0.200 and 10.0. This is true for both our production and repo source build assets. 

When we target only NetCurrent in repo source build then only a NetCurrent compiler, libs, etc ... are flowed to repos still targeting 9.0.200. When they do their repo level source build they are quite possibly only using a `net9.0` SDK, or using `net9.0` only TFMS. That breaks down when we hand them compilers / libraries tied to `net10.0`.

I can't say for certain how that applies to full source build. But have vague memories of us trying to make that NetCurrent pure and failing. 

But it's somewhat immaterial for workspaces layer. That is a component that flows, hence it participates in repo source build and must abide by the NetCurrent + NetPrevious restriction.

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply eng/targets/TargetFrameworks.props:58]: > When we target only NetCurrent in repo source build then only a NetCurrent compiler, libs, etc ... are flowed to repos still targeting 9.0.200. When they do their repo level source build they are quite possibly only using a net9.0 SDK, or using net9.0 only TFMS. That breaks down when we hand them compilers / libraries tied to net10.0.
I can't say for certain how that applies to full source build. But have vague memories of us trying to make that NetCurrent pure and failing.

I fully get the point why the repo source-build needs to target both `NetPrevious` and `NetCurrent`. In full product source-build, `NetCurrent` should be enough as it automatically adapts to the TFM that is required in the given context and release. I.e. in 9.0.200 NetCurrent=net9.0 but when you insert into 10.0.100, NetCurrent=net10.0.

One unrelated observation: it would be great if the repo source-build leg would also target the TFM that we insert into the (latest) product. Right now it depends on the repo SDK (9.0.100)  and target net8.0 and net9.0 (NetPrevious and NetCurrent) with Arcade 9. The latest product though builds with the 10.0.100 SDK and net10.0 TFM. To me it looks like we are missing validation here.

[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: > . In full product source-build, NetCurrent should be enough as it automatically adapts to the TFM that is required in the given context

In programming _should_ is a four letter word. ;)

Again, I feel like every time we try this we convince ourselves it's safe, find out that in practice it's not and we revert. I'm happy to take the bet again but it's more up to you all. 

>  it would be great if the repo source-build leg would also target the TFM t

I'm hopeful that once we get VMR source build working that we can eliminate repo source build entirely. It seems to be of diminishing value given how quickly we get full VMR validation post merge.

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply eng/targets/TargetFrameworks.props:58]: > I'm hopeful that once we get VMR source build working that we can eliminate repo source build entirely. It seems to be of diminishing value given how quickly we get full VMR validation post merge.

We were actually considering the opposite by adding another repo validation leg to mimic the product build (VMR) when not building from source. I'm not sure whether we would prefer to find breaks post-merge during insertion from roslyn -> sdk. I admit that the current (source-build) validation leg doesn't validate all important cases which is why I asked above whether it should target the newer TFM and run with the newer Arcade SDK and .NET SDK.

cc @mmitche @MichaelSimons

[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: My mental model is that we have repo source build because the time between when a change merges and when we get full source build validation is very long. For a normal change the validation is probably ~3 days, and pretty regularly it can week(s). It's also hard to spot test a change because you really need an official build + darc flow to do a source build test. 

None of that is true with VMR source build. The time between merging a change and getting full source build validation is hours. That is because we flow code not binaries. So the moment we merge the flow should happen to VMR. Also as a dev if I think a change is source build risky it's trivial to validate locally. Just apply my change to VMR, run the source build leg (can even put up a PR) and I will get validation. 

At that point I would push back on the need for repo local validation of source build.

[the remark this answers is not in the document]
Comment (ViktorHofer) [reply eng/targets/TargetFrameworks.props:58]: I forwarded your message to our internal communication channel. Thanks for providing your input. Now is the best time to share opinions as we haven't started the repo level validation epic yet. We will discuss this in length (presumably in our next sync meeting) and get back to you.

**Comment by ViktorHofer** [comment]
Can we get at least a workaround merged in sooner than later to unblock https://github.com/dotnet/sdk/pull/45353? Doesn't need to be the perfect solution. Just want to get builds unblocked.

**Comment by jjonescz** [comment]
The current state of the PR should work right? If yes, I just need a sign off to merge it :)

**Comment by ViktorHofer** [comment]
Looks good to me but I'm not a roslyn repo expert...

File: eng/targets/TargetFrameworks.props (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaredpar) [reply eng/targets/TargetFrameworks.props:58]: Happy to come to a sync and elaborate in detail about this.

**Comment by jaredpar** [comment]
> The current state of the PR should work right? 

It works but I'm not sure it's the best. @jasonmalinowski this is really up in your layer. My intuition is that you need consistency in the TFM sets you have in the workspaces layer. At the moment they don't look consistent. If we take this change it's possible this will cause you diff headaches down the road.

**Comment by jasonmalinowski** [comment]
tl;dr: I think this PR is fine as it is, let's merge it.

Longer set of thoughts:

Minimally two things must be using NetRoslynBuildHostNetCoreVersion. This line in the BuildHost:

https://github.com/dotnet/roslyn/blob/7c7927c3ba0aac47641a02798f6273b8ed0a01cc/src/Workspaces/Core/MSBuild.BuildHost/Microsoft.CodeAnalysis.Workspaces.MSBuild.BuildHost.csproj#L7

and this line in the consuming library:

https://github.com/dotnet/roslyn/blob/7c7927c3ba0aac47641a02798f6273b8ed0a01cc/src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj#L101

And right now that's the currently the case. What the actual value is for purposes of source build doesn't really matter much, as long as it's not newer than the SDK where this has to get inserted.

The other place that can cause the missync is we have this version here: 

https://github.com/dotnet/roslyn/blob/7c7927c3ba0aac47641a02798f6273b8ed0a01cc/src/Workspaces/Core/MSBuild/Microsoft.CodeAnalysis.Workspaces.MSBuild.csproj#L7

Which I guess is why this broke originally: we end up with a NetPrevious project with a reference to a NetCurrent since the BuildHost also gets consumed as a library via the project reference. That general mess was being cleaned up in https://github.com/dotnet/roslyn/pull/72257, which I plan to resurrect next month in some form. That'd move the main library to be netstandard and severs some other ProjectReferences so things aren't so coupled. At that point we shouldn't have problems going forward, other than just picking an appropriate value for NetRoslynBuildHostNetCoreVersion, where the logic would generally be:

1. For source build, probably go with NetPrevious because of Jared's point we insert into more than one SDK.
2. For our official build that produces our NuGets, pick whatever is the oldest TFM you can target that doesn't anger your infrastructure folks. And the reason for this is because our MSBuildWorkspace can be used in third-party apps, and it's not uncommon for folks to add them to whatever is the oldest in-support .NET. Or, consider the upgrade assistant case, where they want their app or analysis to be able to run on older runtimes precisely so users can analyze their apps before the upgrade them to something new. So maybe we don't need .NET 5 anymore, but .NET 6 would still be good lest we block people trying to upgrade!

**Comment by mmitche** [comment]
re; NetPrevious vs. NetCurrent. In .NET 9 `NetPrevious` was empty (due to issues in restore when `NetCurrent`== `NetPrevious`). 

```
    <!-- The TFM of the major release of .NET that the Arcade SDK aligns with. -->
    <NetCurrent>net9.0</NetCurrent>

    <!-- The previously released version of .NET.
         Undefined when NetMinimum and NetPrevious are identical. -->
    <NetPrevious />

    <!-- Lowest supported version of .NET at the time of the release of NetCurrent.
         E.g. net8.0 when NetCurrent is net9.0. -->
    <NetMinimum>net8.0</NetMinimum>
```

https://github.com/dotnet/arcade/blob/b1e481cac6e02ff965206de919cd46837181e5a6/src/Microsoft.DotNet.Arcade.Sdk/tools/TargetFrameworkDefaults.props#L12-L21.

I think it is acceptable to use `NetPrevious` here, since this is not inserting into a 9.0.2xx VMR where we would need to source build, **as long as** targeting NetPrevious does not require bringing in the downlevel runtime packs. If that is the case, then roslyn  (and whoever else is forcing roslyn to NetPrevious) must use net10/NetCurrent.

**Review by ViktorHofer — APPROVED** [review]
> tl;dr: I think this PR is fine as it is, let's merge it.

Let's merge it to unblock the arcade build.
